[Merged by Bors] - feat(ring_theory/discriminant): add the discriminant of a family of vectors

riccardobrasca opened the pull request:
We add the definition and some basic results about the discriminant.

From FLT-regular.

- [x] depends on: #10657 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #10353 


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca committed:
add ring_theory/discriminant

riccardobrasca committed:
add fintype.linear_independent_iff''

riccardobrasca committed:
add nondegenerate_iff_det_ne_zero

riccardobrasca committed:
add mul_transpose_eq_reindex_mul_reindex_transpose

riccardobrasca committed:
feat(linear_algebra/bilinear_form):add lemmas

eric-wieser commented on src/data/matrix/basic.lean:
```diff
@@ -1441,6 +1441,12 @@ lemma conj_transpose_reindex [has_star α] (eₘ : m ≃ l) (eₙ : n ≃ o) (M
 
+lemma mul_transpose_eq_reindex_mul_reindex_transpose [fintype n] [fintype m] [semiring α]
+  (e : m ≃ n) (M : matrix n m α) :
+  M ⬝ Mᵀ = (reindex (equiv.refl n) e M) ⬝ (reindex (equiv.refl n) e M)ᵀ :=
```
Is there a reason not to state this as the following instead? `minor` is the simp-normal form.
```suggestion
  M ⬝ Mᵀ = (M.minor id e) ⬝ (M.minor id e)ᵀ :=
```
(note here that `e` is now the original `e.symm`)

eric-wieser committed:
feat(linear_algebra/bilinear_form): add lemmas about congr

riccardobrasca committed:
golf

riccardobrasca committed:
Update src/linear_algebra/bilinear_form.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
Update src/linear_algebra/bilinear_form.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
Update src/linear_algebra/bilinear_form.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
Merge remote-tracking branch 'origin/eric-wieser/bilin_form.congr_lemmas' into RB_bilinear

riccardobrasca committed:
add congr_nondegenerate_iff and golf

riccardobrasca committed:
add nondegenerate_iff_det_ne_zero

riccardobrasca committed:
Merge remote-tracking branch 'origin/RB_bilinear' into RB_discriminant

riccardobrasca committed:
useless assumption

riccardobrasca committed:
Merge remote-tracking branch 'origin/RB_bilinear' into RB_discriminant

riccardobrasca committed:
fix build

eric-wieser committed:
Merge branch 'master' into RB_bilinear

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

riccardobrasca committed:
Update src/linear_algebra/bilinear_form.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
fix build

riccardobrasca committed:
golf

riccardobrasca committed:
Merge remote-tracking branch 'origin/RB_bilinear' into RB_discriminant

riccardobrasca committed:
fix build

riccardobrasca committed:
Update src/linear_algebra/bilinear_form.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Merge remote-tracking branch 'origin/RB_bilinear' into RB_discriminant

eric-wieser commented on src/linear_algebra/linear_independent.lean:
```diff
@@ -146,6 +146,10 @@ theorem fintype.linear_independent_iff' [fintype ι] :
     (linear_map.lsum R (λ i : ι, R) ℕ (λ i, linear_map.id.smul_right (v i))).ker = ⊥ :=
 by simp [fintype.linear_independent_iff, linear_map.ker_eq_bot', funext_iff]
 
+lemma fintype.linear_independent_iff'' [fintype ι] :
```
```suggestion
lemma fintype.linear_dependent_iff [fintype ι] :
```

riccardobrasca committed:
Update src/linear_algebra/linear_independent.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

riccardobrasca committed:
Merge branch 'RB_discriminant' of github.com:leanprover-community/mathlib into RB_discriminant

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+/-- Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define
+`trace_matrix A ι b` as the matrix whose `(i j)`-th element is the trace of `b i * b j`. -/
+noncomputable
+def trace_matrix (b : ι → B) : matrix ι ι A := (λ i j, trace_form A B (b i) (b j))
```
You might find the equation lemmas more convenient with:
```suggestion
def trace_matrix (b : ι → B) : matrix ι ι A
| i j := trace_form A B (b i) (b j)
```

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+def trace_matrix (b : ι → B) : matrix ι ι A
+| i j := trace_form A B (b i) (b j)
+
+@[simp] lemma trace_matrix_apply (b : ι → B) (i j : ι) :
```
You could remove this lemma and just put `@[simp]` on the def above, and the effect would be the same.

riccardobrasca replied:
Done. I've added
```lean
lemma trace_matrix_def (b : ι → B) : trace_matrix A b = λ i j, trace_form A B (b i) (b j) := rfl
```
that is useful in a proof.

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+`trace_matrix A ι b` as the matrix whose `(i j)`-th element is the trace of `b i * b j`. -/
+noncomputable
+def trace_matrix (b : ι → B) : matrix ι ι A
+| i j := trace_form A B (b i) (b j)
```
Another possible spelling would be
```suggestion
:=
bilin_form.to_matrix_aux (trace_form A B) b
```
(Without the bad line wrapping)

riccardobrasca replied:
I think the other version using `| i j := trace_form A B (b i) (b j)` is more clear.

riccardobrasca committed:
fix build

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

jcommelin commented on src/data/matrix/basic.lean:
```diff
@@ -1442,6 +1442,11 @@ lemma conj_transpose_reindex [has_star α] (eₘ : m ≃ l) (eₙ : n ≃ o) (M
   (reindex eₘ eₙ M)ᴴ = (reindex eₙ eₘ Mᴴ) :=
 rfl
 
+lemma mul_transpose_eq_reindex_mul_reindex_transpose [fintype n] [fintype m] [semiring α]
```
The statement uses `minor` instead of `reindex`. So I guess the name should be
```suggestion
lemma mul_transpose_eq_minor_mul_minor_transpose [fintype n] [fintype m] [semiring α]
```

jcommelin commented on src/linear_algebra/linear_independent.lean:
```diff
@@ -146,6 +146,10 @@ theorem fintype.linear_independent_iff' [fintype ι] :
     (linear_map.lsum R (λ i : ι, R) ℕ (λ i, linear_map.id.smul_right (v i))).ker = ⊥ :=
 by simp [fintype.linear_independent_iff, linear_map.ker_eq_bot', funext_iff]
 
+lemma fintype.linear_dependent_iff [fintype ι] :
```
```suggestion
lemma fintype.not_linear_independent_iff [fintype ι] :
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+/-!
+# Discriminant.
+Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define the
+  discriminant of `b` as the determinant of the matrix whose `(i j)`-th element is the trace of
```
```suggestion
*discriminant* of `b` as the determinant of the matrix whose `(i j)`-th element is the trace of
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+
+## Implementation details
+Our definition works for any `A`-algebra `B`, but note that if `B` is not free as an `A`-module,
+  then `trace A B = 0` by definition, so `discriminant A b = 0` for any `b`.
```
```suggestion
then `trace A B = 0` by definition, so `discriminant A b = 0` for any `b`.
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+variable {A}
+
+lemma trace_matrix_of_matrix_vec_mul [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+ trace_matrix A ((P.map (algebra_map A B)).vec_mul b) = Pᵀ ⬝ (trace_matrix A b) ⬝ P :=
```
```suggestion
  trace_matrix A ((P.map (algebra_map A B)).vec_mul b) = Pᵀ ⬝ (trace_matrix A b) ⬝ P :=
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+end
+
+lemma trace_matrix_of_matrix_mul_vec [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+ trace_matrix A ((P.map (algebra_map A B)).mul_vec b) = P ⬝ (trace_matrix A b) ⬝ Pᵀ :=
```
```suggestion
  trace_matrix A ((P.map (algebra_map A B)).mul_vec b) = P ⬝ (trace_matrix A b) ⬝ Pᵀ :=
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+  In this case, in presnce of `h : fintype.card ι = finrank A B`, one can take
+  `e := equiv_of_card_eq ((alg_hom.card A B C).trans h.symm)`. -/
+def embeddings_matrix_reindex (b : ι → B) (e : ι ≃ (B →ₐ[A] C)) :=
+  reindex (equiv.refl ι) e.symm (embeddings_matrix A C b)
```
```suggestion
reindex (equiv.refl ι) e.symm (embeddings_matrix A C b)
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+
+lemma embeddings_matrix_reindex_eq_vandermonde (pb : power_basis A B)
+  (e : fin pb.dim ≃ (B →ₐ[A] C)) : embeddings_matrix_reindex A C pb.basis e =
+  (vandermonde (λ i, e i pb.gen))ᵀ :=
```
```suggestion
  (e : fin pb.dim ≃ (B →ₐ[A] C)) :
  embeddings_matrix_reindex A C pb.basis e = (vandermonde (λ i, e i pb.gen))ᵀ :=
```

jcommelin commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+  (b : ι → B) := by { classical, exact (trace_matrix A b).det }
+
+lemma discriminant_def [decidable_eq ι] [fintype ι] (b : ι → B) :
+  discriminant A b = (trace_matrix A b).det := by convert rfl
```
```suggestion
  discriminant A b = (trace_matrix A b).det := by congr
```

riccardobrasca replied:
`congr` timeouts here.

jcommelin replied:
Oooh, interesting. I thought that they were equivalent.

riccardobrasca committed:
Update src/data/matrix/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/linear_algebra/linear_independent.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
fix build

riccardobrasca committed:
Merge branches 'RB_discriminant' and 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+* `discriminant A b` : the discriminant of `b : ι → B`.
+
+## Main results
+* `zero_of_not_linear_independent` : if `b` is not linear independent, then
```
I know doc-gen doesn't care, but normally we add blank lines after headings:
```suggestion
# Discriminant

Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define the
*discriminant* of `b` as the determinant of the matrix whose `(i j)`-th element is the trace of
`b i * b j`.

## Main definition

* `discriminant A b` : the discriminant of `b : ι → B`.

## Main results

* `zero_of_not_linear_independent` : if `b` is not linear independent, then
```

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+  corresponding to `j : ι` via a bijection `e : ι ≃ (L →ₐ[K] E)`.
+* `of_power_basis_eq_prod` : the discriminant of a power basis.
+
+## Implementation details
```
```suggestion
## Implementation details

```

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+  ext α β,
+  rw [trace_matrix, vec_mul, dot_product, vec_mul, dot_product, mul_apply, bilin_form.sum_left,
+    fintype.sum_congr _ _ (λ (i : ι), @bilin_form.sum_right _ _ _ _ _ _ _ _
+    (b i * P.map (algebra_map A B) i α) (λ (y : ι), b y * P.map (algebra_map A B) y β)), sum_comm],
```
Can you avoid the `@` if you use a `simp_rw` here? I haven't tried.

riccardobrasca replied:
It doesn't seem to work
```lean
simplify tactic failed to simplify
```

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+/-- `embeddings_matrix A C b : matrix ι (B →ₐ[A] C) C` is the matrix whose `(i, σ)` coefficient is
+  `σ (b i)`. It is mostly useful for fields when `fintype.card ι = finrank A B` and `C` is
+  algebraically closed. -/
+def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C := (λ i (σ : B →ₐ[A] C), σ (b i))
```
```suggestion
@[simp] def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C
| i σ := σ (b i))
```
might be better, as that results in `rw embeddings_matrix` only working when it has indices, which prevents matrices decaying into functions in unexpected ways.

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,221 @@
+lemma discriminant_def [decidable_eq ι] [fintype ι] (b : ι → B) :
+  discriminant A b = (trace_matrix A b).det := by convert rfl
+
+namespace discriminant
```
Putting this in the `discriminant` namespace seems slightly unusual; we don't put lemmas about `matrix.det` under a `det` namespace. So maybe `discriminant_zero_of_not_linear_independent`? Maybe it's worth renaming `discriminant` to `discr` to keep things concise?

riccardobrasca committed:
use discr_ instead of discriminant.

riccardobrasca committed:
rename discriminant to discr

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
add blank lines after headings

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

riccardobrasca committed:
fix build

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+
+## Main definition
+
+* `discr A b` : the discriminant of `b : ι → B`.
```
```suggestion
* `algebra.discr A b` : the discriminant of `b : ι → B`.
```
since this is in the `algebra` namespace, right?

eric-wieser replied:
Same for the "main results" below

eric-wieser commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+  (e : ι ≃ (L →ₐ[K] E)) : (trace_matrix K b).map (algebra_map K E) =
+  (embeddings_matrix_reindex K E b e) ⬝ (embeddings_matrix_reindex K E b e)ᵀ :=
+by rw [trace_matrix_eq_embeddings_matrix_mul_trans, embeddings_matrix_reindex, reindex_apply,
+  mul_transpose_eq_minor_mul_minor_transpose, ← equiv.coe_refl, equiv.refl_symm]
```
Do you still need the `root`? Aren't you already in the `algebra` namespace now?

riccardobrasca committed:
root not needed

riccardobrasca committed:
fix docstring

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

Vierkantor commented on src/data/matrix/basic.lean:
```diff
@@ -1442,6 +1442,11 @@ lemma conj_transpose_reindex [has_star α] (eₘ : m ≃ l) (eₙ : n ≃ o) (M
 
+lemma mul_transpose_eq_minor_mul_minor_transpose [fintype n] [fintype m] [semiring α]
+  (e : n ≃ m) (M : matrix n m α) :
+  M ⬝ Mᵀ = (M.minor id e) ⬝ (M.minor id e)ᵀ  :=
```
I could see this as a `simp` lemma if it was defined the other way, or is there a reason that the original direction is preferred?
```suggestion
@[simp]
lemma mul_transpose_eq_minor_mul_minor_transpose [fintype n] [fintype m] [semiring α]
  (e : n ≃ m) (M : matrix n m α) :
  (M.minor id e) ⬝ (M.minor id e)ᵀ = M ⬝ Mᵀ :=
```

riccardobrasca replied:
Ah, it's the simp normal form linter that is not happy about this. I will have a closer look later.

Vierkantor commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+/-- `embeddings_matrix_reindex A C b e : matrix ι ι C` is the matrix whose `(i, j)` coefficient
+  is `σⱼ (b i)`, where `σⱼ : B →ₐ[A] C` is the embedding corresponding to `j : ι` given by a
+  bijection `e : ι ≃ (B →ₐ[A] C)`. It is mostly useful for fields and `C` is algebraically closed.
+  In this case, in presnce of `h : fintype.card ι = finrank A B`, one can take
```
```suggestion
  In this case, in presence of `h : fintype.card ι = finrank A B`, one can take
```

Vierkantor commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,222 @@
+  bijection `e : ι ≃ (B →ₐ[A] C)`. It is mostly useful for fields and `C` is algebraically closed.
+  In this case, in presnce of `h : fintype.card ι = finrank A B`, one can take
+  `e := equiv_of_card_eq ((alg_hom.card A B C).trans h.symm)`. -/
+def embeddings_matrix_reindex (b : ι → B) (e : ι ≃ (B →ₐ[A] C)) :=
```
I recall defining this matrix (and the previous ones) for the proof of [`det_trace_form_ne_zero`](https://github.com/leanprover-community/mathlib/blob/master/src/ring_theory/trace.lean#L348), but ended up inlining it apparently. It's probably worth the time to move `embeddings_matrix_reindex` and some of its lemmas to `trace.lean` and seeing if we can use them to simplify the proofs there.

If you don't have the time for that right now, I'll see if I can do that later today.

riccardobrasca replied:
I don't have time today to do it, but if you are busy feel free to suggest here what you think it should be done and I will take care of it.

Vierkantor replied:
I'm happy to do it now, while the code is still fresh in my mind. Do you mind if I push directly to your branch?

riccardobrasca replied:
No problem!

Vierkantor replied:
Done!

riccardobrasca committed:
Update src/data/matrix/basic.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

riccardobrasca committed:
fix build

Vierkantor committed:
Use `trace_matrix` and `embeddings_matrix` in `trace.lean`

Vierkantor committed:
Add docstrings

riccardobrasca committed:
this isn't needed

Vierkantor committed:
Lint (+ name) fix

eric-wieser commented on src/data/matrix/basic.lean:
```diff
@@ -1442,6 +1442,12 @@ lemma conj_transpose_reindex [has_star α] (eₘ : m ≃ l) (eₙ : n ≃ o) (M
+lemma minor_mul_transpose_minor [fintype n] [fintype m] [semiring α]
+  (e : n ≃ m) (M : matrix n m α) :
+  (M.minor id e) ⬝ (Mᵀ).minor e id = M ⬝ Mᵀ :=
+by rw [←minor_mul_equiv, minor_id_id]
```
Is this evidence that `minor_mul_equiv` should be stated in reverse and marked simp?

riccardobrasca replied:
Let me see if it breaks a lot of things.

riccardobrasca replied:
I don't know if stating `minor_mul_equiv` in reverse is better or not, but I think that in any case it is something for another PR, since one should modify also `reindex_linear_equiv_mul` and so on.

eric-wieser replied:
Sorry, I didn't mean that as a suggestion for _this_ PR!

Vierkantor replied:
I think reversing `minor_mul_equiv` is probably a good idea. I've opened #10638 so we don't forget.

riccardobrasca committed:
fix build

Vierkantor reviewed (approved):
LGTM. Since I wrote a bit of code, I'll let @jcommelin decide.

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

alexjbest commented on src/linear_algebra/linear_independent.lean:
```diff
@@ -146,6 +146,10 @@ theorem fintype.linear_independent_iff' [fintype ι] :
     (linear_map.lsum R (λ i : ι, R) ℕ (λ i, linear_map.id.smul_right (v i))).ker = ⊥ :=
 by simp [fintype.linear_independent_iff, linear_map.ker_eq_bot', funext_iff]
 
+lemma fintype.not_linear_independent_iff [fintype ι] :
```
The corresponding theorem on line 125 is called `linear_dependent_iff`, it would be good to keep these names consistent.

riccardobrasca replied:
looking at the statement `not_linear_independent_iff` seems a better name. Let me see if it's easy to change line 125.

alexjbest replied:
Yes I do agree, I would also prefer that they say not linear independent I think, but have no idea if that would be a big change to make and distract from this PR

riccardobrasca replied:
It seems a quite small change.

alexjbest commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,139 @@
+  `algebra.discr A ((P.map (algebra_map A B)).mul_vec b)`.
+* `algebra.discr_not_zero_of_linear_independent` : over a field, if b` is linear independent, then
+  `algebra.discr K b ≠ 0`.
+* `algebra.discr_eq_det_embeddings_matrix_reindex_pow_two` if `L/K` is a fields extension and
```
```suggestion
* `algebra.discr_eq_det_embeddings_matrix_reindex_pow_two` if `L/K` is a field extension and
```

alexjbest commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,139 @@
+* `algebra.discr_of_matrix_vec_mul` and `discr_of_matrix_mul_vec` : formulas relating
+  `algebra.discr A ι b` with `algebra.discr A ((P.map (algebra_map A B)).vec_mul b)` and
+  `algebra.discr A ((P.map (algebra_map A B)).mul_vec b)`.
+* `algebra.discr_not_zero_of_linear_independent` : over a field, if b` is linear independent, then
```
```suggestion
* `algebra.discr_not_zero_of_linear_independent` : over a field, if `b` is linear independent, then
```

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Alex J Best <alex.j.best@gmail.com>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Alex J Best <alex.j.best@gmail.com>

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

YaelDillies commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,139 @@
+import linear_algebra.vandermonde
+
+/-!
+# Discriminant.
```
```suggestion
# Discriminant
```
Given that there are lots of different discriminants, maybe it's better to explicit it a bit?
```suggestion
# Discriminant of a family of vectors
```
or are all discriminants somehow unifiable under a common framework?

riccardobrasca committed:
better name

riccardobrasca committed:
fix build

riccardobrasca committed:
fix build

Vierkantor committed:
refactor(data/matrix): reverse the direction of `matrix.minor_mul_equiv`

In #10350 this change was proposed, since we apparently use that backwards way more than we use it forwards.

We also change `reindex_linear_equiv_mul`, which is similarly much more popular backwards than forwards.

Closes: #10638

riccardobrasca committed:
Merge remote-tracking branch 'origin/reverse_minor_mul_equiv' into RB_discriminant

riccardobrasca committed:
fix build

Vierkantor committed:
Don't need these underscores any more!

riccardobrasca committed:
Merge remote-tracking branch 'origin/reverse_minor_mul_equiv' into RB_discriminant

YaelDillies commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,139 @@
+import linear_algebra.matrix.determinant
+import ring_theory.trace
+import ring_theory.norm
+import linear_algebra.vandermonde
```
One import fits them all
```suggestion
import ring_theory.trace
```

YaelDillies commented on src/ring_theory/discriminant.lean:
```diff
@@ -0,0 +1,139 @@
+by rw [discr_def, ring_hom.map_det, ring_hom.map_matrix_apply,
+    trace_matrix_eq_embeddings_matrix_reindex_mul_trans, det_mul, det_transpose, pow_two]
+
+lemma discr_of_power_basis_eq_prod (e : fin pb.dim ≃ (L →ₐ[K] E)) :
```
Why this lemma name?

riccardobrasca replied:
It computes the discriminant of a `power_basis` , that is a product of something. I don't see a reasonably short name that includes also the description of the product, but it's a quite well known result, so saying `product` should be enough.

jcommelin replied:
```suggestion
lemma discr_power_basis_eq_prod (e : fin pb.dim ≃ (L →ₐ[K] E)) :
```

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

riccardobrasca committed:
doc

riccardobrasca committed:
Merge branch 'master' of github.com:leanprover-community/mathlib

riccardobrasca committed:
Merge branch 'master' into RB_discriminant

jcommelin reviewed (commented):
I think it's best to remove `of_` from the `discr_of_power_basis_eq_prod` name. Otherwise, let's get this in mathlib!

bors d+

riccardobrasca committed:
Update src/ring_theory/discriminant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
